pandas/util/_decorators.py

```diff
@@ -268,17 +270,28 @@ def decorator(func: F) -> F:
         def wrapper(*args, **kwargs) -> Callable:
             return func(*args, **kwargs)
 
-        templates = [func.__doc__ if func.__doc__ else ""]
+        # collecting docstring and docstring templates
+        docstring_components: List[Union[str, Callable]] = []
+        if func.__doc__:
+            docstring_components.append(dedent(func.__doc__))
+
         for arg in args:
-            if isinstance(arg, str):
-                templates.append(arg)
-            elif hasattr(arg, "_docstr_template"):
-                templates.append(arg._docstr_template)  # type: ignore
-            elif arg.__doc__:
-                templates.append(arg.__doc__)
-
-        wrapper._docstr_template = "".join(dedent(t) for t in templates)  # type: ignore
-        wrapper.__doc__ = wrapper._docstr_template.format(**kwargs)  # type: ignore
+            if hasattr(arg, "_docstring_components"):
+                docstring_components.extend(arg._docstring_components)  # type: ignore
```

Comment:
Not sure if this could be a better fix for this one:
```suggestion
            if hasattr(arg, "_docstring_components") and isinstance(arg._docstring_components, list):
                docstring_components.extend(arg._docstring_components)
```

Reply:
Hi @datapythonista. I would consider this as two parts. One is to add type check for `_docstring_components` and one for remove `# type: ignore`.

#### For remove `# type: ignore` 
After I made the change, I still get the errors below and not sure if we can remove that.
```bash
pandas/util/_decorators.py:280: error: Item "str" of "Union[str, Callable[..., Any]]" has no attribute "_docstring_components"
pandas/util/_decorators.py:280: error: Item "function" of "Union[str, Callable[..., Any]]" has no attribute "_docstring_components"
```

*I guess my settings are not the same as checks here. I even the last passed build, I still get mypy errors on local. So, the change might fix it and just I don't know.*

#### For checking if `_docstring_components` is `list`
I think this is a very good point, we do as defensive programming. But I am a little bit curious about how far we should go to protect it. Also, since this function is internal use for pandas only, would it be better to let it failed? Then the developer would get notified there is a conflict immediately?

Just to be clarified, I would be happy to make this change if you (or anyone) feels needed. It is just I didn't see a clear improvement on this, because I see both pros and cons for this.

Reply:
I'm happy to merge this as is, was wondering if the suggested change would fix the problem.

We can surely go back to it in the future if needed, but let's merge this first, since this already adds a lot of value, and we can start replacing `Appender` by `doc` at a larger scale after this.